src/sql-parser/src/lexer/buf.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::parser::ParserError;
+
+/// A cursor over a string with a variety of lexing convenience methods.
+pub struct LexBuf<'a> {
+    buf: &'a str,
+    pos: usize,
+}
+
+impl<'a> LexBuf<'a> {
+    /// Creates a new lexical buffer.
+    ///
+    /// The internal cursor is initialized to point at the start of `buf`.
+    pub fn new(buf: &'a str) -> LexBuf<'a> {
+        LexBuf { buf, pos: 0 }
+    }
+
+    /// Returns the next character in the buffer, if any, without advancing the
+    /// internal cursor.
+    pub fn peek(&self) -> Option<char> {
+        self.buf[self.pos..].chars().next()
```

Comment:
This looks really slow. Is it actually ok?

Reply:
Yeah, it's ok. It compiles down quite a bit: https://godbolt.org/z/aTq4o9

It's not nothing, but not obviously more expensive than any Unicode decoder. Certainly not more expensive then all the dumb strings we're cloning everywhere in the parser.